File: src/main/java/com/parmet/squashlambdas/activity/Court.java
package com.parmet.squashlambdas.activity;

import static com.google.common.base.CaseFormat.UPPER_CAMEL;
import static com.google.common.base.CaseFormat.UPPER_UNDERSCORE;
import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static com.parmet.squashlambdas.activity.Sport.HARDBALL;
import static com.parmet.squashlambdas.activity.Sport.RACQUETS;
import static com.parmet.squashlambdas.activity.Sport.SQUASH;
import static com.parmet.squashlambdas.activity.Sport.TENNIS;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public enum Court {
  COURT_1(SQUASH),
  COURT_2(SQUASH),
  COURT_3(SQUASH),
  COURT_5(HARDBALL),
  COURT_6(HARDBALL),
  COURT_7(HARDBALL),
  TENNIS_COURT(TENNIS),
  RACQUETS_COURT(RACQUETS);

  // "Court: Court #x" (match creation)
  // "Courts: Court #x" (activity creation)
  private static final Pattern COURT = Pattern.compile(".*Courts?: Court #(\\d) [/\\-].*");

  private final Sport sport;
  private final String pretty =
      UPPER_UNDERSCORE.to(UPPER_CAMEL, name().replaceAll("_", " "));

  private Court(Sport sport) {
    this.sport = checkNotNull(sport, "sport");
  }

  /** e.g., "Tennis & Racquet Club / Court: Court #7" */
  public static Court fromLocationString(String body) {
    Matcher m = COURT.matcher(body);
    checkArgument(m.matches(), "Unable to parse court from %s", body);
    return valueOf("COURT_" + Integer.parseInt(m.group(1)));
  }

  public Sport getSport() {
    return sport;
  }

  @Override
  public String toString() {
    return pretty;
  }
}

File: src/main/kotlin/com/parmet/squashlambdas/activity/Court.kt
package com.parmet.squashlambdas.activity

import com.parmet.squashlambdas.activity.Sport.Fitness
import com.parmet.squashlambdas.activity.Sport.Hardball
import com.parmet.squashlambdas.activity.Sport.Racquets
import com.parmet.squashlambdas.activity.Sport.Squash
import com.parmet.squashlambdas.activity.Sport.Tennis
import com.parmet.squashlambdas.notify.CourtSerializer
import kotlinx.serialization.Serializable
import java.util.regex.Pattern

@Serializable(with = CourtSerializer::class)
enum class Court(
    val sport: Sport,
    val pretty: String
) {
    Court1(Squash, "Court 1"),
    Court2(Squash, "Court 2"),
    Court3(Squash, "Court 3"),
    Court5(Hardball, "Court 5"),
    Court6(Hardball, "Court 6"),
    Court7(Hardball, "Court 7"),
    TennisCourt(Tennis, "Tennis Court"),
    RacquetsCourt(Racquets, "Racquets Court"),
    FitnessClasses(Fitness, "Fitness Classes");

    override fun toString() =
        pretty

    companion object
}

// "Court: Court #x" (match creation)
// "Courts: Court #x" (activity creation)
private val NUMBERED_COURT = Pattern.compile(".*Court #(\\d) [/\\-] (Squash|Hardball).*")

// "Court: Court Tennis - Court Tennis" (match creation)
// "Court: Court Tennis / Court Tennis" (player joins)
private val TENNIS_COURT = Pattern.compile(".*Court Tennis [-/] Court Tennis.*")

// "Court: Racquets - Racquets" (match creation)
// "Court: Racquets / Racquets" (match deletion)
private val RACQUETS_COURT = Pattern.compile(".*Racquets [-/] Racquets.*")

private val courtsByPretty = Court.entries.associateBy { it.pretty }

internal fun Court.Companion.fromPrettyName(value: String) =
    requireNotNull(courtsByPretty[value]) {
        "No court named $value"
    }

internal fun Court.Companion.fromLocationString(body: String) =
    when {
        TENNIS_COURT.matcher(body).matches() ->
            Court.TennisCourt

        RACQUETS_COURT.matcher(body).matches() ->
            Court.RacquetsCourt

        else -> {
            val matcher = NUMBERED_COURT.matcher(body)
            require(matcher.matches()) { "Unable to parse court from $body" }
            fromPrettyName("Court ${matcher.group(1)}")
        }
    }
